### PR TITLE
Add initial coding style

### PR DESCRIPTION
**Pull request by berewt**
The PR is at the moment a WIP, we need to discuss its content before merging it.

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Formatting, PNG optimization, etc. are updated
    - [ ] Important changes are reflected in changelog.d of the affected packages
- PR
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reference the ADR in the PR and reference the PR in the ADR (if revelant)
    - [ ] Reviewer requested

### CODE REVIEW

File: STYLEGUIDE.adoc [hunk 1]
```diff
@@ -0,0 +1,578 @@
...
+data Term tyname name uni fun a
+----
+
+==== Reasoning
+
+The reasoning is essentially the same as for type signatures. We've got used to GHC inferring all this, and in the past we didn't even have the means to easily state kind signatures. But these days with people using fancier type-level machinery, and with better support for kind signatures, it's time to just start writing them down.
+
+=== Write Haddock
+
+Every top-level exported binding should have Haddock. Non-exported bindings should probably have Haddock too.
```
Comment (kayvank) [inline STYLEGUIDE.adoc:166]: :100:

File: STYLEGUIDE.adoc [hunk 2]
```diff
@@ -0,0 +1,578 @@
...
+
+==== The Good
+
+These are basically all fine and can be put in `default-extensions`.
+
+- Anything in https://github.com/ghc-proposals/ghc-proposals/pull/380[`GHC2021`]. Once we have a GHC version that supports the GHC2021 language, we will likely switch to using it.
+- `LambdaCase`: clear, helpful
+- `DerivingStrategies`: always
+- `GADTs`: well established, useful
+- `OverloadedStrings`: essential when working with `Text`, which you should
```
Comment (kayvank) [inline STYLEGUIDE.adoc:404]: we should add this to marconi cabal files

File: STYLEGUIDE.adoc [hunk 3]
```diff
@@ -0,0 +1,578 @@
...
+- `NegativeLiterals`
+- `DerivingVia`: great
+- `RoleAnnotations`: if you need it, you need it
+
+==== The Situational
+
+The following extensions are generally fine if you find that they’ll make your life much easier, but you probably don’t want to use them *all* the time.
+
+- `RecordWildCards`
```
Comment (kayvank) [inline STYLEGUIDE.adoc:413]: I am OK with this one in Test modules only

File: STYLEGUIDE.adoc [hunk 4]
```diff
@@ -0,0 +1,578 @@
...
+
+[source,Haskell]
+----
+sortWithOrder :: Ord a => SortOrder -> [a] -> [a]
+sortWithOrder Ascending  = id . sort
+sortWithOrder Descending = reverse . sort
+----
+
+===== Instead of destructor functions
+
```
Comment (berewt) [inline STYLEGUIDE.adoc:205]: Remove this point

File: STYLEGUIDE.adoc [hunk 5]
```diff
@@ -0,0 +1,578 @@
...
+[source,Haskell]
+----
+fmap (either f g) eithers
+----
+
+==== Reasoning
+
+Pattern matching is easy to read, and allows the compiler to give better errors and warnings (e.g. incomplete match warnings).
+
+=== Avoid punning
```
Comment (berewt) [inline STYLEGUIDE.adoc:233]: Remove it.

File: STYLEGUIDE.adoc [hunk 6]
```diff
@@ -0,0 +1,578 @@
...
+Good:
+
+[source,Haskell]
+----
+data Foo = MkFoo Int
+----
+
+=== Avoid ticks in function names
+
+Generally avoid using ticks to distinguish function names. All this conveys is that it is “another” version of the function. Try expressing the difference in the function name, even if it makes it longer.
```
Comment (berewt) [inline STYLEGUIDE.adoc:254]: It's okay locally (in a where/let) clauses

File: STYLEGUIDE.adoc [hunk 7]
```diff
@@ -0,0 +1,578 @@
...
+Therefore if you need to produce output for humans, define a `Pretty` instance.
+This will typically need to be hand-written, unless it happens that you can defer to the `Show` instance, e.g. for simple enums `Show` can be fine since it just prints the constructor names.
+
+=== Use the package name as the root of the module name hierarchy
+
+If the package is `foo-bar`, then the modules should all be `FooBar.X`.
+
+==== Reasoning
+
+See “Naming conventions” https://www.haskellforall.com/2021/05/module-organization-guidelines-for.html[here]. We do it slightly differently (“FooBar” rather than “Foo.Bar”), but the main principle is the same.
```
Comment (berewt) [inline STYLEGUIDE.adoc:347]: Both options are fine.

File: STYLEGUIDE.adoc [hunk 8]
```diff
@@ -0,0 +1,578 @@
...
+[source,Haskell]
+----
+import Control.Lens (first, … , _Right) hiding (ix, lens)
+----
+
+Good:
+
+[source,Haskell]
+----
+import Control.Lens qualified as Lens
```
Comment (berewt) [inline STYLEGUIDE.adoc:447]: We can still use a separate import for operators.

File: STYLEGUIDE.adoc [hunk 9]
```diff
@@ -0,0 +1,578 @@
...
+
+Not using deriving strategies requires the compiler to guess which strategy you want. This can have consequences, especially when `DeriveAnyClass` is enabled, since you can accidentally end up using anyclass deriving when you didn’t mean to. Better to be explicit.
+
+=== Use Text
+
+Use Text instead of String unless you have a good reason not to.
+
+==== Reasoning
+
+It’s 2022, use a proper, unicode aware string type instead of a linked list.
```
Comment (james-iohk) [inline STYLEGUIDE.adoc:322]: Perhaps remove year reference, I'm sure the point remains relevant :)

File: STYLEGUIDE.adoc [hunk 10]
```diff
@@ -0,0 +1,578 @@
...
+
+=== data-default
+
+Don’t use `data-default`, instead just define `defaultX` values for your `X` type.
+
+==== Reasoning
+
+`Default` is not terribly bad. It’s truly ad-hoc polymorphism: all you get is name reuse, you can’t (or shouldn’t) write a function that’s polymorphic over `Default a`. That’s fine, but it also means that the benefit is fairly marginal.
+
+Additionally, just defining specific values is more flexible. If, say, you want multiple default values for different contexts, then that is straightforward with values but not with `Default`.
```
Comment (berewt) [inline STYLEGUIDE.adoc:504]: Drop this one.

File: STYLEGUIDE.adoc [hunk 11]
```diff
@@ -0,0 +1,578 @@
...
+=== Make data strict by default, use `nothunks` for long-lived data structures
+
+Use `StrictData` for new code; make fields strict unless you have a good reason not to.
+
+For data structures that might live for a long time, use `nothunks` to assert that they don't contain unexpected thunks.
+
+==== Reasoning
+
+A painful lesson of Haskell in production is that space leaks really suck, are a huge pain to track down, and can originate in surprising locations.
+This suggests that it's worth a bit of proactive paranoia: just make things strict as much as possible, in the hopes of squashing any nascent space leaks.
```
Comment (koslambrou) [inline STYLEGUIDE.adoc:390]: nascent?

File: STYLEGUIDE.adoc [hunk 12]
```diff
@@ -0,0 +1,578 @@
+= Plutus Haskell style guide
```
Comment (james-iohk) [inline STYLEGUIDE.adoc:1]: "Plutus Apps" or "Plutus Tools"?

Comment (berewt) [reply STYLEGUIDE.adoc:1]: Yes it was shamelessly copied/pasted from plutus-core, I guess "repo-name Style guide" should be fine?

File: STYLEGUIDE.adoc [hunk 13]
```diff
@@ -0,0 +1,578 @@
...
+
+[source,Haskell]
+----
+fmap (either f g) eithers
+----
+
+==== Reasoning
+
+Pattern matching is easy to read, and allows the compiler to give better errors and warnings (e.g. incomplete match warnings).
+
```
[the remark this answers is not in the document]
Comment (berewt) [reply STYLEGUIDE.adoc:232]: Maybe not needed as it's a standard linting rule.

File: STYLEGUIDE.adoc (hunk 11 above)
Comment (berewt) [reply STYLEGUIDE.adoc:390]: shorter than "before it appears" and can be understood from the context, no?

Comment (koslambrou) [reply STYLEGUIDE.adoc:353]: Ah, I thought that word was a typo. I didn't know the meaning.

File: STYLEGUIDE.adoc [hunk 14]
```diff
@@ -0,0 +1,578 @@
...
+Good:
+
+[source,Haskell]
+----
+f x = y
+  where
+    y = ...
+----
+
+Avoid lines over 100 characters. This is codified in `.editorconfig`.
```
[the remark this answers is not in the document]
Comment (berewt) [reply STYLEGUIDE.adoc:41]: We decided to keep 10 as a compromise

**Comment by berewt** [comment]
I've updated it based on the feedback of our technical meeting, let me know if you think I missed anything.
(again, no addition at this stage, just removing/adapting stuff to our need).

File: STYLEGUIDE.adoc [hunk 15]
```diff
@@ -0,0 +1,543 @@
...
+
+==== Reasoning
+
+The indentation is arbitrary. The line length is longer than the “traditional” 80 characters,
+but anecdotally Haskell code tends toward longer lines, and breaking them up doesn’t help readability much.
+
+Placing the double-colon on the same line makes it easier to find definitions.
+We don't necessarily have HLS running all the time, and it may not work on every module.
+
+`fourmolu` is a fairly modest formatter that doesn’t touch too much. At the moment we mostly use it for formatting imports!
```
Comment (koslambrou) [inline STYLEGUIDE.adoc:98]: Delete, that was only relevant for `stylish-haskell` :D

**Pull request merged**